edelphi/src/main/java/fi/metatavu/edelphi/utils/QueryDataUtils.java

```diff
@@ -139,6 +176,147 @@ public int compare(QueryPage o1, QueryPage o2) {
+          });
+        }
+      }
+    }
+    
+    return rows;
+  }
+  
+  /**
+   * Writes a CSV file
+   * 
+   * @param columnHeaders headers
+   * @param rows rows
+   * @return CSV file
+   * @throws IOException throws IOException when CSV writing fails
+   */
+  private static byte[] writeCsv(String[] columnHeaders, List<String[]> rows) throws IOException {
+    return writeCsv(columnHeaders, rows.stream().toArray(String[][]::new));
+  }
+
+  /**
+   * Writes a CSV file
+   * 
+   * @param columnHeaders headers
+   * @param rows rows
+   * @return CSV file
+   * @throws IOException throws IOException when CSV writing fails
+   */
+  private static byte[] writeCsv(String[] columnHeaders, String[][] rows) throws IOException {
+    try (ByteArrayOutputStream csvStream = new ByteArrayOutputStream(); OutputStreamWriter streamWriter = new OutputStreamWriter(csvStream, Charset.forName("UTF-8"))) {
+      CSVWriter csvWriter = new CSVWriter(streamWriter, ',');
+
+      csvWriter.writeNext(columnHeaders);
+      
+      for (String[] row : rows) {
+        csvWriter.writeNext(row);
+      }
+
+      csvWriter.close();
+      
+      return csvStream.toByteArray();
+    }
+  }
+  
+  /**
+   * Returns label for given replier export strategy
+   * 
+   * @param locale locale
+   * @param replierExportStrategy replier export strategy
+   * @return label
+   */
+  private static String getReplierExportStrategyLabel(Locale locale, ReplierExportStrategy replierExportStrategy) {
+    switch (replierExportStrategy) {
+      case NONE:
+      break;
+      case HASH:
+        return Messages.getInstance().getText(locale, "panelAdmin.query.export.csvReplierIdColumn");
+      case NAME:
+        return Messages.getInstance().getText(locale, "panelAdmin.query.export.csvReplierNameColumn");
+      case EMAIL:
+        return Messages.getInstance().getText(locale, "panelAdmin.query.export.csvReplierEmailColumn");
+    }
+    
+    return null;
+  }
+  
+  /**
+   * Returns user identifier for given replier export strategy
+   * 
+   * @param replierExportStrategy replier export strategy
+   * @param queryReply reply
+   * @return user identifier
+   */
+  private static String getReplierExportStrategyValue(ReplierExportStrategy replierExportStrategy, QueryReply queryReply) {
+    if (queryReply != null) {
+      User user = queryReply.getUser();
+      if (user != null) {
+        switch (replierExportStrategy) {
+          case NONE:
+          break;
+          case HASH:
+            return RequestUtils.md5EncodeString(String.valueOf(user.getId()));
+          case NAME:
+            return user.getFullName(true, false);
+          case EMAIL:
+            return user.getDefaultEmailAsString();
+        }
+      }
+    }
+    
+    return "-";
+  }
   
   private static byte[] exportDataToCsv(Locale locale, ReplierExportStrategy replierExportStrategy, List<String> columns, Map<QueryReply, Map<Integer, Object>> rows) throws IOException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)